### PR TITLE
Bugfix/rgcn

### PR DESCRIPTION
**Pull request by kieranricardo**
This pull request fixes the issues with running RGCN in eager mode with `num_bases > 0`.  In the case of `num_bases > 0` the `relational_kernels` are now calculated in `RelationalGraphConvolution.call` instead of `RelationalGraphConvolution.build`.  It appears that all operations with `tensorflow.Variable` must occur in the call function in a `keras.layer` if eager execution is used.

Reviewer checklist:

- [ ] All unit tests pass
- [ ] The demo notebook runs correctly
- [ ] The results of the demo are reproducible (excluding random effects)
- [ ] The demo notebook does not give gradient warnings

### CODE REVIEW

File: stellargraph/layer/rgcn.py [hunk 1]
```diff
@@ -322,8 +316,14 @@ def call(self, inputs):
         # Calculate the layer operation of RGCN
         output = K.dot(features, self.self_kernel)
 
+        if self.num_bases > 0:
+            self.relational_kernels = [
```
Comment (huonw) [inline stellargraph/layer/rgcn.py]: It's a little weird to me that this is mutating the `self` object, and saving these computed values outside this `call` method. In particular, if doing this computation has to happen inside the `call` method, it might not be valid to use the results after the `call` method, and storing them in `self` is risking hitting that problem.

As such, maybe:

```python
if self.num_bases > 0:
    relational_kernels = [ ... ]
else:
    relational_kernels = self.relational_kernels

for i in range(self.num_relationships):
    h_graph = K.dot(...)
    output += K.dot(..., relational_kernels[I])
```

But... possibly my suspicion is wrong and it's totally fine and even useful to hold the result for longer. If that's the case, then maybe we could reduce how much work we do here (assuming `call` is called more than once) by not recomputing the kernels:

```python
if self.relational_kernels is None:
    assert self.num_bases > 0, "delayed initialisation only works for basis-matrix formulation"
    self.relational_kernels = [ ... ]
```

File: demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb [hunk 2]
```diff
@@ -186,16 +176,16 @@
...
-   "execution_count": 6,
+   "execution_count": 16,
    "metadata": {},
    "outputs": [
     {
      "name": "stderr",
      "output_type": "stream",
      "text": [
-      "/Users/wil9dw/projects/stellargraph/stellargraph-4/stellargraph/core/experimental.py:88: ExperimentalWarning: RGCN is experimental: it has severe known bugs (see: https://github.com/stellargraph/stellargraph/issues/649, https://github.com/stellargraph/stellargraph/issues/677). It may be difficult to use and may have major changes at any time.\n",
+      "/Users/kieranricardo/PycharmProjects/origin-stellargaph/stellargraph/stellargraph/core/experimental.py:88: ExperimentalWarning: RGCN is experimental: it has severe known bugs (see: https://github.com/stellargraph/stellargraph/issues/649, https://github.com/stellargraph/stellargraph/issues/677). It may be difficult to use and may have major changes at any time.\n",
```
Comment (huonw) [inline demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: Not related to this PR, but... did you know you've spelled stellargraph wrong in this directory name? 😝 

```suggestion
      "/Users/kieranricardo/PycharmProjects/origin-stellargraph/stellargraph/stellargraph/core/experimental.py:88: ExperimentalWarning: RGCN is experimental: it has severe known bugs (see: https://github.com/stellargraph/stellargraph/issues/649, https://github.com/stellargraph/stellargraph/issues/677). It may be difficult to use and may have major changes at any time.\n",
```

File: demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb [hunk 3]
```diff
@@ -312,8 +318,8 @@
      "text": [
       "\n",
       "Test Set Metrics:\n",
-      "\tloss: 1.3316\n",
-      "\tacc: 0.5556\n"
+      "\tloss: 0.4581\n",
+      "\tacc: 0.8889\n"
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: Seems to be a result random train/test split - with the previous run to this one (same code) I got 94.8% accuracy. The paper only uses one train/test split

File: stellargraph/layer/rgcn.py [hunk 4]
```diff
@@ -252,12 +252,6 @@ def build(self, input_shapes):
                 for _ in range(self.num_relationships)
             ]
 
-            # create a kernel for each edge type from a linear combination
-            # of the basis matrices
-            self.relational_kernels = [
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply stellargraph/layer/rgcn.py]: yeah good point, leaving an object partially initialized is weird

File: stellargraph/layer/rgcn.py (hunk 1 above)
Comment (kieranricardo) [reply stellargraph/layer/rgcn.py]: I tried this but it yielded unconnected gradient errors (weirdly different to the prior warnings). `self.relational_kernels` is the product of trainable variables so it has to be recalculated every training step. 

My suspicion is that in non-eager mode this can be done implicitly by baking this calculation in the computational graph but in eager mode this recalculation has to done be explicitly. Anyway, I'll try and neaten this sectino up a bit

File: demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb (hunk 2 above)
Comment (kieranricardo) [reply demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: Nothing gets past your reviews! I'm hopeless without spell check 😝

File: stellargraph/layer/rgcn.py (hunk 1 above)
Comment (huonw) [reply stellargraph/layer/rgcn.py]: > I tried this

My comment had two suggestions, so I'm not sure what "this" means. From context it sounds like the second suggestion? In which case, it does sound like using the results of these computations outside `call` is incorrect and so I would recommend trying something along the lines of the first suggestion.

File: demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: I take that back! increasing the number of parameters yields consistently better test results 94% +

[the remark this answers is not in the document]
Comment (huonw) [reply demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: Ah, so you think that this is within the natural variance of the algorithm due to weight initialisation (since there's no other randomness in this algorithm, right?)?

> The paper only uses one train/test split

Are you referring to the train/test split from [1]? (That is, http://data.dws.informatik.uni-mannheim.de/rmlod/LOD_ML_Datasets/data/old/datasets/AIFB/) It looks like we're not using the same split?

Also, the paper says:

- "averaged over 10 runs": do we get similar accuracy averaged over 10 runs?
- "We further set aside 20% of the training set as a validation set for hyperparameter tuning": it seems like this demo still uses 80% for the training set?

[1] Ristoski, P.; de Vries, G. K. D.; and Paulheim, H. 2016. A collection of benchmark datasets for systematic evaluations of machine learning on the semantic web. In International Semantic Web Conference, 186–194. Springer

File: demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb [hunk 5]
```diff
@@ -227,7 +217,23 @@
...
+     "text": [
+      "Epoch 1/20\n",
+      "1/1 [==============================] - 17s 17s/step - loss: 1.5992 - acc: 0.4296 - val_loss: 1.5667 - val_acc: 0.4167\n",
+      "Epoch 2/20\n",
+      "1/1 [==============================] - 17s 17s/step - loss: 1.5456 - acc: 0.5141 - val_loss: 1.5178 - val_acc: 0.4167\n",
+      "Epoch 3/20\n",
+      "1/1 [==============================] - 17s 17s/step - loss: 1.4708 - acc: 0.5634 - val_loss: 1.4537 - val_acc: 0.4167\n",
+      "Epoch 4/20\n",
+      "1/1 [==============================] - 17s 17s/step - loss: 1.4482 - acc: 0.5141 - val_loss: 1.3817 - val_acc: 0.3889\n",
+      "Epoch 5/20\n"
```
[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: the output should now be deleted - it had a warning msg with personal details

File: demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (kieranricardo) [reply demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb]: With the initial parameters I had the natural weight + train test variance affected the final results significantly. It looks like with a larger number of parameters (layer sizes of 32 and 20 bases) the variance of final results decreases and we see results consistently comparable to the papers results.

And yeah, we're using a random split instead of their split, mostly because the single file dataset I downloaded was easier to handle than the one linked (but doesn't contain the train/test/val split).

File: stellargraph/layer/rgcn.py (hunk 1 above)
Comment (kieranricardo) [reply stellargraph/layer/rgcn.py]: ah sorry, I was referring to your second suggestion. I've implemented your first suggestion which really tidies up the call function.
